examples/widgets/search_entry.nim

```diff
@@ -0,0 +1,88 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+# 
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE
+
+import owlkettle, owlkettle/[playground, adw, dataentries]
+import std/[strutils, sequtils]
+
+viewable App:
+  searchDelay: uint = 100
+  text: string = ""
+  placeholderText: string = "Search List"
+  sensitive: bool = true
+  tooltip: string = ""
+  sizeRequest: tuple[x, y: int] = (-1, -1)
+  
+  items: seq[string] = mapIt(0..<100, "Search Entry " & $it)
+  filteredItems: seq[string] = mapIt(0..<100, "Search Entry " & $it)
+  selected: int = 0
+  
+method view(app: AppState): Widget =
+  let isInActiveSearch = app.text != ""
+  result = gui:
+    Window():
+      defaultSize = (600, 400)
+      HeaderBar() {.addTitlebar.}:
+        insert(app.toAutoFormMenu(ignoreFields = @["filteredItems"])) {.addRight.}
+      
+        SearchEntry() {.addTitle, expand: true.}:
+          margin = Margin(top:0, left: 48, bottom:0, right: 48)
+          text = app.text
+          searchDelay = app.searchDelay
+          placeholderText = app.placeholderText
+          sensitive = app.sensitive
+          tooltip = app.tooltip
+          sizeRequest = app.sizeRequest
+          
+          proc nextMatch() = 
+            if app.selected < app.filteredItems.high:
+              app.selected += 1
+          
+          proc previousMatch() = 
+            if app.selected > 0:
+              app.selected -= 1
+          
+          proc searchStarted(searchString: string) = 
+            echo "Search Started: ", searchString
+          
+          proc activate(searchString: string) =
+            echo "activated search for entry: ": $app.filteredItems[app.selected]
+          
+          proc changed(searchString: string) = 
+            app.text = searchString
+            app.selected = 0
+            app.filteredItems = app.items.filterIt(searchString in it)
+          
+          proc stopSearch(searchString: string) = 
+            echo "Search Stopped"
+            app.text = ""
+            app.filteredItems = app.items
+          
+      ScrolledWindow:
+        Box(orient = OrientY):
```

Comment:
This should be a `ListBox` which allows for selection (2 way bind `app.selected` to `ListBox.selected`). The labels probably look nicer when left aligned + ~6 pixel margin. In case of the ListBox I would also remove the `<b>` for the selected item and use ListBox.selected instead.

Reply:
Done, it's listBox and I used a Box to be able to align left since I can't set hAlign within an adder to ListBox (ListBox doesn't take in seq[BoxChild], but rather seq[Widget]).
```nim
      ScrolledWindow:
        ListBox:
          selected = [app.selected].toHashSet()
          selectionMode = SelectionSingle
          
          proc select(rows: Hashset[int]) =
            for num in rows:
              app.selected = num
          
          for index, item in app.filteredItems:
            Box():
              Label(text = item, margin = 6) {.hAlign: AlignStart, expand: false.}
        
```